src/action/client_side_file_finder/glob_to_regex.rs

```diff
@@ -0,0 +1,86 @@
+// Copyright 2020 Google LLC
+//
+// Use of this source code is governed by an MIT-style license that can be found
+// in the LICENSE file or at https://opensource.org/licenses/MIT.
+
+use crate::session::RegexParseError;
+use regex::Regex;
+
+/// Converts glob expression into a Rust Regex.
+/// E.g. "asd*[!123]??" will be converted into "asd.*[^123]..".
+/// This implementation is a Rust port of the cpython code:
+/// https://github.com/python/cpython/blob/2.7/Lib/fnmatch.py
+fn glob_to_regex(pat: &str) -> Result<Regex, RegexParseError> {
```

Comment:
I am not sure if I like this. In Python we use `fnmatch` to translate glob to a regex and then use this for matching because it avoids implementing globbing mechanism ourselves. However, if we have to implement `fnmatch` ourselves, it defeats the whole purpose of its existence and we can implement globbing directly, without unnecessarily going through a regex.

Reply:
In RRG we have to implement globbing ourselves. Part of globbing implementation is deciding if a directory/file name matches a pattern. I see 2 ways of doing that:
1) Convert the glob to regex::Regex and use the regex library to check if name matches the pattern.
2) Implement a functionality `matches_glob(text, glob) -> bool`.

I believe that difficulty of both solutions is similar (because of edge-cases like unmatched parenthesis, characters escaping etc). In this PR I decided to implement conversion to regex as we had a working implementation from cpython that was already used in GRR and I don't expect it to cause any problems.

Reply:
> Part of globbing implementation is deciding if a directory/file name matches a pattern. 

This is just the way GRR does it and it is not the most efficient of doing things. The proper way would be to split path into proper glob components and do traversal and matching ourselves. I understand that the solution presented here is a sort of acceptable temporary hack and maybe I am willing to accept it, but this is definitely not the way things should be implemented in a production-ready RRG agent.

Reply:
Also, have you verified all the licensing implications? In GRR was just copy-and-pasted from the standard library without any changes. Here, I am not sure what are the legal requirements.

Reply:
This implementation was only inspired by the CPython code - I changed the comment to reflect it.

I don't agree that using Regex can't be a very efficient way of doing things. The plan is to do the traversal ourselves (you can see a working version here: https://github.com/Spawek/rrg/blob/dev/src/action/finder/action.rs - it's not ready for PR yet) and use Regex just to match directory names.